### DISCUSSION TITLE
Is sub-query support available ?

### DISCUSSION THREAD

**Discussion by blazing-gig** [collerek/ormar#227] [Q&A]
Hey @collerek ,

Great work on this library. I have one question though. When I read this doc on [join and sub-queries](https://collerek.github.io/ormar/queries/joins-and-subqueries/), I was only able to find examples of how to perform joins. I was expecting to see some examples on sub-queries as well. 

Is support currently available for this ? If yes, could you please point me to the relevant section in the docs for the same ?

**Comment by collerek** [comment]
Hi, yeah the title might be a bit misleading as Ormar automatically creates subqueries for you i.e. during pagination/limit/offset to paginate on real models and not database rows. 

What exactly would you like to do, as custom queries (like select something as column) is not supported out of the box. What is your desired subquery?

**Reply by blazing-gig** [reply]
He @collerek ,

Thanks for responding.

By SubQuery, I meant something like [this](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#subquery-expressions)

**Reply by collerek** [reply]
Ok, so the problem with this approach is that you define an additional column, which as of now you cannot add to ormar model, without defining the field upfront. 

You can explicitly set `Config.extra="allow"` in pydantic, but you cannot do this in `ormar`. That might be an enhancement proposal, but any extra field is by definition not validated, and I'm not yet sure if i like this idea, as if it's not validated you cannot fully rely on this field. 

Maybe better would be to define something like subquery field in a way that is similar to generated column described in #196.

As of now in order to support subquery you would have to return the data as raw data (list of dicts or tuples) that skips validation. This is proposed in issue #223, that I'm working on right now.

With 223 finished i was thinking about allowing to execute arbitrary queries too, kind of something similar to django annotations.

The actual subquery support is another enhancement, question is if this should be handled as some kind of field definition with expected data type that can be validated. If yes it should also probably be read only.

**Reply by blazing-gig** [reply]
Hey @collerek ,

I think we've touched base on a few things here. The gist I believe is that currently support for something similar to Django annotations would be added in the future. 

Regd the question of whether it would have to be added as another field which would go through proper type validation, I'm not so sure since these fields are alive only for the duration of the QuerySet and are completely ad-hoc. But if required, something like this could be implemented:

```py
class AnnotationWrapper:
    expression: Any = 'the expr to evaluate'
    output_type: Any = some_python_type

...

A.objects.annotate(some_adhoc_field=AnnotationWrapper(some_expr, output_type=some_python_type))
```
Actually something like the above makes much more sense only if there were separate expressions defining aggregation operations like `add`, `count` etc., and something like [F expressions](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#f-expressions) to access fields directly at a SQL level.

Are there ideas for implementing something similar to F expressions and [expression wrappers](https://docs.djangoproject.com/en/3.2/ref/models/expressions/#expressionwrapper-expressions) because they seem to be important and useful constructs to have as part of the library. 

Let me know your thoughts on the same.

**Reply by collerek** [reply]
Yeah the F expressions and wrapper looks interesting and definitely worth implementing
